src/aspire/source/simulation.py

```diff
@@ -468,3 +475,132 @@ def eval_coords(self, mean_vol, eig_vols, coords_est):
+           Defaults to mean via `estimate_signal_mean_energy`.
+           Can use variance method via `estimate_signal_var`.
+        :returns: Simulation source.
+        """
+
+        # Create a Simulation
+        sim = cls(*args, **kwargs)
+
+        # When a user supplies angles and requests `sample_n` subset of images
+        #   randomize the rotations for them to avoid as much bias as possible.
+        # The original rotations stored in `_rots` will be restored later.
+        shuffle = (not sim._uniform_random_angles) and (sample_n is not None)
+        if shuffle:
+            # Store the original rotations.
+            _rots = sim.rotations.copy()
+            # Shuffle the Sim rotations
+            sim.rotations = sim.rotations[np.random.permutation(sim.n)]
+
+        # Assert NoiseAdder has not been provided
+        if "noise_adder" in kwargs or sim.noise_adder is not None:
+            raise RuntimeError(
+                "Cannot provide 'noise_adder' when using {cls.__name__}.from_snr."
+            )
+
+        # Estimate the required noise variance
+        signal_var = sim.estimate_signal_power(
+            sample_n=sample_n,
+            support_radius=support_radius,
+            signal_power_method=signal_power_method,
+        )
+        noise_var = signal_var / target_snr
+
+        # Assign the noise_adder
+        sim.noise_adder = WhiteNoiseAdder(var=noise_var)
+        logger.info(f"Appended {sim.noise_adder} to generation pipeline")
+
+        # Restore the original rotations when previously shuffled.
+        if shuffle:
+            sim.rotations = _rots
+
+        return sim
+
+    def estimate_psnr(self, sample_n=None, batch_size=512, units=None):
+        """
+        Estimate Peak SNR as max(signal)^2 / MSE,
+        where MSE is computed between `projections`
+        and the resulting simulated `images`.
+        PSNR is computed along the stack axis and an average
+        value across the stack sample is returned.
+        Note that PSNR is inherently a poor metric for identical images.
+
+        :param sample_n: Number of images used for estimate.
+            Defaults to all images in source.
+        :param units: Optionally, convert from default ratio to log scale (`dB`).
+        :returns: Estimated peak signal to noise ratio.
+        """
+
+        if sample_n is None:
+            sample_n = self.n
+
+        if sample_n > self.n:
+            logger.warning(
+                f"`estimate_psnr` sample_n > Source.n: {sample_n} > {self.n}."
+                f" Accuracy may be impaired, settting sample_n=self.n={self.n}"
+            )
+            sample_n = self.n
+
+        peaksq = np.empty(sample_n, dtype=self.dtype)
+        mse = np.empty(sample_n, dtype=self.dtype)
+        for start in trange(0, sample_n, batch_size):
+            end = min(start + batch_size, sample_n)
+
+            signal = self.projections[start:end].asnumpy()
```

Comment:
Yes it will.  I tried both and this way seemed to work slightly better for me. There is also the matter of shifts and amplitude corruptions too. This preempts them all.  I  chose to take the straight projections after considering that to be the best possible representation for true signal.

It is possible in the future we will just move to some 3d thing that has been mentioned, but I need to see it first... that would also be before any of those corruptions anyway....

Keep in mind the "noise" is computed from the ears of the image, with "signal" coming from the center.  Adding corruption to the image, like a lot of zero centered noise, will increase its variance. If this was added into the power computation, we would be overestimated the signal strength. (I think).

What MSE do you mean?

Reply:
Ok sounds good. I figured it was something you had thought about already. I was speaking of the MSE use in the calculation below (lines 593 - 597).